### ISSUE TITLE
feat: Provide way to change tag data on Feast objects (e.g. entities & feature views)

### ISSUE THREAD

**Issue by tchughesiv** [feast-dev/feast#4459]
**Is your feature request related to a problem? Please describe.**
https://github.com/feast-dev/feast/issues/2337

**Describe the solution you'd like**
We should be able to tag objects directly, similar to `kubectl label` for example. With the initial implementation we should be able to do the following:
1. add new labels
2. overwrite existing labels
3. remove labels

For the following object types:
- DataSource
- Entity
- FeatureService
- FeatureView
- OnDemandFeatureView
- SavedDatasetSpec
- StreamFeatureView
- ValidationReference

**Additional context**
I'm thinking something like the following for cli:
```shell
$ feast tag --help
Usage: feast tag [OPTIONS] OBJECT_TYPE NAME

  [Experimental] Tag objects

Options:
  --tags TEXT  Tags (e.g. --tags 'key:value' --tags 'key:value, key:value,
               ...').  [required]
  --overwrite  Overwrite tag(s)
  --help       Show this message and exit.
```
```shell
$ feast tag entity driver --tags release:production
```

**Comment by tchughesiv** [comment]
@tokoko now worries... but couldn't what you just described regarding `feast apply` actually be an argument FOR this kind of `feast tag` capability. currently, how does one update feast object tags without re-instantiating everything? as first mentioned here -
https://github.com/feast-dev/feast/issues/2337

that said, i'll leave the decision to the broader feast community.

**Comment by franciscojavierarceo** [comment]
As mentioned in #2337 
>I'd expect feast apply to make those changes in the registry, but not for example tear down / re-instantiate an online store

I think this approach is reasonable. The alternative approach to get the same result is to modify the apply to have tag specific logic in the apply for tags. 

This is a good PR with a lot of changes incorporated to support this, so I'm inclined to include this as is. We can note revisiting it in the future if the imperative use causes problems.

**Comment by tchughesiv** [comment]
@tokoko @franciscojavierarceo i'm completely fine with pivoting here and implementing whatever is best. what would be a better design for this?
 - we could change the command `feast tag` to mimic `feast apply` but instead only tag objects?
 - we could modify `feast apply` by adding a `--tags-only` or `--tags` option to indicate only update tags.
 - something else?

**Comment by tokoko** [comment]
@tchughesiv that one might be a bit of a misunderstanding. I don't think #2337 is referring to the problem that this PR is trying to solve. Instead the problem described there is that any update of a feature view (even if only tags are updated) invokes an update call for the feature view in question in `OnlineStore` [here](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/infra/online_stores/online_store.py#L287). Even that's almost like a non-problem if you ask me as in most online store implementations that method is either empty or idempotent, so extra pointless calls really don't make any difference.

> but couldn't what you just described regarding `feast apply` actually be an argument FOR this kind of `feast tag` capability? currently, how does one update feast object tags without re-instantiating everything?

That's fair, but the problem here is that feast currently pushes for the workflow in which all the resources in a feast project is described in a git repo in a python file which is then applied to the registry all at once. Were we to introduce any sort of alternative flow to update any individual resources, all those changes would simply be reverted the next time a user runs `feast apply`. That's my main concern here... unless a user also manually changes those tags in the original python definitions, any changes made will be short-lived and confusing.

btw, I'm not the biggest fan of `feast apply`'s all-or-nothing approach either, but imho the right way to tackle that is to introduce a way to allow users to run `apply` on a subset of resources (#4199), not to allow updating specific attributes.

> we could modify `feast apply` by adding a `--tags-only` or `--tags` option to indicate only tag updates.

I don't really see the point, tbh. You normally need to manually make changes to python definitions before running apply. If you want to update tags only, you would simply edit only the tags in the original file(s). If the goal is to protect against mistakes a `feast plan` command is the better solution (sort of like `terraform plan` and `terraform apply` workflow).

> something else?

The only flow that I think could work if we really need this feature is to differentiate between tags that are added with `apply` and tags that are added in an imperative way. Maybe have a different field for the latter in protos or come up with some sort of naming convention where tags with keys that look like this `manual_*` are to be left alone by a `feast apply` command. idk, I guess that's doable, but even then I personally don't think there's an actual use case for this so far.

> whatever we decide... would we want to keep the tag key/value regex validation i've added in the existing PR or not?

Can you elaborate on this one? That's a really scary-looking regex :smile: What's the downside of allowing all strings w/o any sort of validation?

**Comment by tchughesiv** [comment]
> Can you elaborate on this one? That's a really scary-looking regex 😄 What's the downside of allowing all strings w/o any sort of validation?
haha, the regex mostly just implementing the same reqs as kubernetes does for labels.
https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set

i will say that the regex validation is much more important to the PR in its current state. if we don't implement it, and there's no appetite for it... again, i have no qualms with just closing the PR.

so far, the only "downside" i've experienced having no tag guardrails is that, during my testing the registry would stop responding to "get" requests when a tag of significant length was added to an object. but i'm sure that's not limited to tags... i expect the same issues might arise with object names, etc.

**Comment by tchughesiv** [comment]
@tokoko @franciscojavierarceo jsut let me know if we should close it out and move forward with other work

**Comment by tokoko** [comment]
> just let me know if we should close it out and move forward with other work

yup, I think we should close this one, btw fully realizing that this thread should have happened 2 weeks ago... my bad.

> so far, the only "downside" i've experienced having no tag guardrails is that, during my testing the registry would stop responding to "get" requests when a tag of significant length was added to an object.

That sounds weird actually. Probably should have a separate ticket for it.

**Comment by tchughesiv** [comment]
> yup, I think we should close this one, btw fully realizing that this thread should have happened 2 weeks ago... my bad.

all good, i prob should have given the issue more time to bake before starting work.

**Comment by tchughesiv** [comment]
@tokoko after doing some testing, it appears there's an issue where tag keys over 1023 characters cause issues when attempting to then "describe" an item. only keys though, extreme value lengths don't appear to cause the same issue.

seems yaml spec sets a 1024 limit on keys. let me know if i should open an issue or not.
```bash
% feast feature-views describe driver_hourly_stats
Traceback (most recent call last):
  File "/Users/tohughes/workspace/venv/bin/feast", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/Users/tohughes/workspace/venv/lib/python3.12/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tohughes/workspace/venv/lib/python3.12/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/Users/tohughes/workspace/venv/lib/python3.12/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tohughes/workspace/venv/lib/python3.12/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tohughes/workspace/venv/lib/python3.12/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tohughes/workspace/venv/lib/python3.12/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tohughes/workspace/venv/lib/python3.12/site-packages/click/decorators.py", line 33, in new_func
    return f(get_current_context(), *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tohughes/workspace/feast/sdk/python/feast/cli.py", line 458, in feature_view_describe
    yaml.safe_load(str(feature_view)), default_flow_style=False, sort_keys=False
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tohughes/workspace/venv/lib/python3.12/site-packages/yaml/__init__.py", line 125, in safe_load
    return load(stream, SafeLoader)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tohughes/workspace/venv/lib/python3.12/site-packages/yaml/__init__.py", line 81, in load
    return loader.get_single_data()
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tohughes/workspace/venv/lib/python3.12/site-packages/yaml/constructor.py", line 49, in get_single_data
    node = self.get_single_node()
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tohughes/workspace/venv/lib/python3.12/site-packages/yaml/composer.py", line 36, in get_single_node
    document = self.compose_document()
               ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tohughes/workspace/venv/lib/python3.12/site-packages/yaml/composer.py", line 55, in compose_document
    node = self.compose_node(None, None)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tohughes/workspace/venv/lib/python3.12/site-packages/yaml/composer.py", line 84, in compose_node
    node = self.compose_mapping_node(anchor)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tohughes/workspace/venv/lib/python3.12/site-packages/yaml/composer.py", line 133, in compose_mapping_node
    item_value = self.compose_node(node, item_key)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tohughes/workspace/venv/lib/python3.12/site-packages/yaml/composer.py", line 84, in compose_node
    node = self.compose_mapping_node(anchor)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tohughes/workspace/venv/lib/python3.12/site-packages/yaml/composer.py", line 133, in compose_mapping_node
    item_value = self.compose_node(node, item_key)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tohughes/workspace/venv/lib/python3.12/site-packages/yaml/composer.py", line 84, in compose_node
    node = self.compose_mapping_node(anchor)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tohughes/workspace/venv/lib/python3.12/site-packages/yaml/composer.py", line 127, in compose_mapping_node
    while not self.check_event(MappingEndEvent):
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tohughes/workspace/venv/lib/python3.12/site-packages/yaml/parser.py", line 98, in check_event
    self.current_event = self.state()
                         ^^^^^^^^^^^^
  File "/Users/tohughes/workspace/venv/lib/python3.12/site-packages/yaml/parser.py", line 549, in parse_flow_mapping_key
    raise ParserError("while parsing a flow mapping", self.marks[-1],
yaml.parser.ParserError: while parsing a flow mapping
  in "<unicode string>", line 22, column 13:
        "tags": {
                ^
expected ',' or '}', but got ':'
  in "<unicode string>", line 23, column 1032:
     ... aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa": ""
```
